services/elasticache/pkg/resource/replication_group/custom_update_api.go

```diff
@@ -490,6 +555,193 @@ func (rm *resourceManager) newUpdateShardConfigurationRequestPayload(
 	return res, nil
 }
 
+// describeCacheCluster provides CacheCluster object
+// per the supplied latest Replication Group Id
+// it invokes DescribeCacheClusters API to do so
+func (rm *resourceManager) describeCacheCluster(
+	ctx context.Context,
+	latest *resource,
+) (*svcsdk.CacheCluster, error) {
+	input := &svcsdk.DescribeCacheClustersInput{}
+
+	var cacheClusterId *string = nil
+	latestStatus := latest.ko.Status
+	if latestStatus.NodeGroups == nil {
+		return nil, nil
+	}
+	for _, nodeGroup := range latestStatus.NodeGroups {
+		if nodeGroup.NodeGroupMembers == nil {
+			continue
+		}
+		for _, nodeGroupMember := range nodeGroup.NodeGroupMembers {
+			if nodeGroupMember.CacheClusterID == nil {
+				continue
+			}
+			cacheClusterId = nodeGroupMember.CacheClusterID
+			break
+		}
+		if cacheClusterId != nil {
+			break
+		}
+	}
+
+	if cacheClusterId == nil {
+		return nil, nil
+	}
+
+	input.SetCacheClusterId(*cacheClusterId)
+	resp, respErr := rm.sdkapi.DescribeCacheClustersWithContext(ctx, input)
+	rm.metrics.RecordAPICall("READ_MANY", "DescribeCacheClusters", respErr)
+	if respErr != nil {
+		rm.log.Error(respErr, "Error during DescribeCacheClusters")
+		return nil, respErr
+	}
+	if resp.CacheClusters == nil {
+		return nil, nil
+	}
+
+	for _, cc := range resp.CacheClusters {
+		if cc == nil {
+			continue
+		}
+		return cc, nil
+	}
+	return nil, nil
+}
+
+// securityGroupIdsDiffer return true if
+// Security Group Ids differ between desired spec and latest (from cache cluster) status
+func (rm *resourceManager) securityGroupIdsDiffer(
+	desired *resource,
+	latest *resource,
+	latestCacheCluster *svcsdk.CacheCluster,
+) bool {
+	if desired.ko.Spec.SecurityGroupIDs == nil {
+		return false
+	}
+
+	desiredIds := []*string{}
+	for _, id := range desired.ko.Spec.SecurityGroupIDs {
+		if id == nil {
+			continue
+		}
+		var value string
+		value = *id
+		desiredIds = append(desiredIds, &value)
+	}
+	sort.Slice(desiredIds, func(i, j int) bool {
+		return *desiredIds[i] < *desiredIds[j]
+	})
+
+	latestIds := []*string{}
+	if latestCacheCluster.SecurityGroups != nil {
```

Comment:
This needs a nil-guard. `latestCacheCluster` can be `nil`.

Reply:
done